### ISSUE TITLE
odbc extension: buffer size & incompatible pointer types

### ISSUE THREAD

**Issue by vpaesa** [little-brother/sqlite-gui#76]
Hi,
Thanks for your SQLite3 extensions.
I compiled them with [MinGW64 ](http://mingw-w64.org/)(you are using the similar but different [MinGW32](https://osdn.net/projects/mingw/)).

The compiler complained on _stprintf() because [Microsoft changed its definition to align it to ISO some time ago](https://stackoverflow.com/questions/594799/how-can-i-use-stprintf-in-my-programs-with-and-without-unicode-support) and now is asking too for the buffer size.
```
$ /usr/bin/x86_64-w64-mingw32-gcc -Os -I ../sqlite-gui/include -shared ../sqlite-gui/extensions/odbc.c -o mingw64-x86_64/odbc.dll -lodbc32 -s -static-libgcc
../sqlite-gui/extensions/odbc.c: In function ‘odbc_read’:
../sqlite-gui/extensions/odbc.c:195:23: warning: passing argument 2 of ‘swprintf’ makes integer from pointer without a cast [-Wint-conversion]
  195 |   _stprintf(create16, TEXT("create table if not exists temp.\"%ls%i\" (\""), tablename16, sid);
      |                       ^~~~
      |                       |
      |                       short unsigned int *
In file included from /usr/x86_64-w64-mingw32/sys-root/mingw/include/stdio.h:1372,
                 from ../sqlite-gui/extensions/odbc.c:42:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/swprintf.inl:34:41: note: expected ‘size_t’ {aka ‘long long unsigned int’} but argument is of type ‘short unsigned int *’
   34 | int swprintf (wchar_t *__stream, size_t __count, const wchar_t *__format, ...)
      |                                  ~~~~~~~^~~~~~~
```
It could be somewhere in your Code:Block environment you have set the old _stprintf() behaviour ([maybe this way?](https://stackoverflow.com/questions/1293889/microsoft-stprintf-warning))

There were also a few warnings on incompatible pointer types:
```
$ /usr/bin/x86_64-w64-mingw32-gcc -Os -I ../sqlite-gui/include -shared ../sqlite-gui/extensions/odbc.c -o mingw64-x86_64/odbc.dll -l odbc32 -s -static-libgcc
../sqlite-gui/extensions/odbc.c: In function ‘odbc_write’:
../sqlite-gui/extensions/odbc.c:413:87: warning: passing argument 6 of ‘SQLBindCol’ from incompatible pointer type [-Wincompatible-pointer-types]
  413 |                         SQLBindCol(hStmt, 4, SQL_C_WCHAR, colName, MAX_COLUMN_LENGTH, &cbColName);
      |                                                                                       ^~~~~~~~~~
      |                                                                                       |
      |                                                                                       SQLINTEGER * {aka long int *}
In file included from /usr/x86_64-w64-mingw32/sys-root/mingw/include/sqlext.h:9,
                 from ../sqlite-gui/extensions/odbc.c:46:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/sql.h:489:157: note: expected ‘SQLLEN *’ {aka ‘long long int *’} but argument is of type ‘SQLINTEGER *’ {aka ‘long int *’}
  489 |   SQLRETURN SQL_API SQLBindCol(SQLHSTMT StatementHandle,SQLUSMALLINT ColumnNumber,SQLSMALLINT TargetType,SQLPOINTER TargetValue,SQLLEN BufferLength,SQLLEN *StrLen_or_Ind);
      |                                                                                                                                                     ~~~~~~~~^~~~~~~~~~~~~
../sqlite-gui/extensions/odbc.c:414:74: warning: passing argument 6 of ‘SQLBindCol’ from incompatible pointer type [-Wincompatible-pointer-types]
  414 |                         SQLBindCol(hStmt, 5, SQL_C_SSHORT, &dataType, 0, &cbDataType);
      |                                                                          ^~~~~~~~~~~
      |                                                                          |
      |                                                                          SQLINTEGER * {aka long int *}
In file included from /usr/x86_64-w64-mingw32/sys-root/mingw/include/sqlext.h:9,
                 from ../sqlite-gui/extensions/odbc.c:46:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/sql.h:489:157: note: expected ‘SQLLEN *’ {aka ‘long long int *’} but argument is of type ‘SQLINTEGER *’ {aka ‘long int *’}
  489 |   SQLRETURN SQL_API SQLBindCol(SQLHSTMT StatementHandle,SQLUSMALLINT ColumnNumber,SQLSMALLINT TargetType,SQLPOINTER TargetValue,SQLLEN BufferLength,SQLLEN *StrLen_or_Ind);
      |                                                                                                                                                     ~~~~~~~~^~~~~~~~~~~~~
../sqlite-gui/extensions/odbc.c:467:158: warning: passing argument 10 of ‘SQLBindParameter’ from incompatible pointer type [-Wincompatible-pointer-types]
  467 |                                                 int rc = isNULL ? SQLBindParameter(hStmt, colNo, SQL_PARAM_INPUT, SQL_C_NUMERIC, SQL_NUMERIC, 0, 0, NULL, 0, &cbNull) :
      |                                                                                                                                                              ^~~~~~~
      |                                                                                                                                                              |
      |                                                                                                                                                              SQLINTEGER * {aka long int *}
In file included from ../sqlite-gui/extensions/odbc.c:46:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/sqlext.h:1442:217: note: expected ‘SQLLEN *’ {aka ‘long long int *’} but argument is of type ‘SQLINTEGER *’ {aka ‘long int *’}
 1442 |   SQLRETURN SQL_API SQLBindParameter(SQLHSTMT hstmt,SQLUSMALLINT ipar,SQLSMALLINT fParamType,SQLSMALLINT fCType,SQLSMALLINT fSqlType,SQLULEN cbColDef,SQLSMALLINT ibScale,SQLPOINTER rgbValue,SQLLEN cbValueMax,SQLLEN *pcbValue);
      |                                                                                                                                                                                                                 ~~~~~~~~^~~~~~~~
```

The attached diff [odbc_arraysize_sqllen.diff.txt](https://github.com/little-brother/sqlite-gui/files/7056285/odbc_arraysize_sqllen.diff.txt) fixed this.
If you think it worth, please consider applying it to odbc.c

**Comment by vpaesa** [comment]
I dowloaded https://github.com/sqlitebrowser/sqlitebrowser/releases/download/v3.12.2/DB.Browser.for.SQLite-3.12.2-win64.zip
Then I put the odbc64.dll you compiled into DB4S extensions folder.
I launched DB4S and set it to use an in-memory database.
Then I used the GUI menu "Tools->Load extension ...", and I got the message "Extension successfully loaded.".
Then I used "Execute SQL" with command
```select odbc_read('Driver={Microsoft Access Text Driver (*.txt, *.csv)};Dbq=c:/vpg/', 'select * from sample.csv', 'sample_csv');```
And it worked fine:
```
Execution finished without errors.
Result: 1 rows returned in 693ms
At line 1:
select odbc_read('Driver={Microsoft Access Text Driver (*.txt, *.csv)};Dbq=c:/vpg/', 'select * from sample.csv', 'sample_csv');
```

Then I exited DB4S, and I copied the 64-bit odbc.dll I compiled (this [recipe](https://github.com/vpaesa/xlsxsqlite/blob/master/compile.sh)) into DB4S extensions folder.
Same steps, and same successful result.

Also success with the "util", "test" extensions whose code is at [error loading the extension library on windows 10](https://github.com/sqlitebrowser/sqlitebrowser/issues/2795)

My test computer runs Microsoft Windows 10 Home

Did you use by chance the .msi release intead of the .zip one in your DB4S tests?
